File: src/main/java/link/infra/packwiz/installer/metadata/ModFile.java
package link.infra.packwiz.installer.metadata;

import com.google.gson.annotations.SerializedName;
import link.infra.packwiz.installer.UpdateManager.Options.Side;
import link.infra.packwiz.installer.metadata.hash.Hash;
import link.infra.packwiz.installer.metadata.hash.HashUtils;
import link.infra.packwiz.installer.request.HandlerManager;
import okio.Source;

import java.util.Map;

public class ModFile {
	public String name;
	public String filename;
	public Side side;

	public Download download;
	public static class Download {
		public SpaceSafeURI url;
		@SerializedName("hash-format")
		public String hashFormat;
		public String hash;
	}

	public Map<String, Object> update;

	public Option option;
	public static class Option {
		public boolean optional;
		public String description;
		@SerializedName("default")
		public boolean defaultValue;
	}

	public Source getSource(SpaceSafeURI baseLoc) throws Exception {
		if (download == null) {
			throw new Exception("Metadata file doesn't have download");
		}
		if (download.url == null) {
			throw new Exception("Metadata file doesn't have a download URI");
		}
		SpaceSafeURI newLoc = HandlerManager.getNewLoc(baseLoc, download.url);
		if (newLoc == null) {
			throw new Exception("Metadata file URI is invalid");
		}
		
		return HandlerManager.getFileSource(newLoc);
	}

	public Hash getHash() throws Exception {
		if (download == null) {
			throw new Exception("Metadata file doesn't have download");
		}
		if (download.hash == null) {
			throw new Exception("Metadata file doesn't have a hash");
		}
		if (download.hashFormat == null) {
			throw new Exception("Metadata file doesn't have a hash format");
		}
		return HashUtils.getHash(download.hashFormat, download.hash);
	}

	public boolean isOptional() {
		if (option != null) {
			return option.optional;
		}
		return false;
	}

}

File: src/main/kotlin/link/infra/packwiz/installer/metadata/ModFile.kt
package link.infra.packwiz.installer.metadata

import cc.ekblad.toml.delegate
import cc.ekblad.toml.model.TomlValue
import cc.ekblad.toml.tomlMapper
import link.infra.packwiz.installer.metadata.curseforge.UpdateData
import link.infra.packwiz.installer.metadata.hash.Hash
import link.infra.packwiz.installer.metadata.hash.HashFormat
import link.infra.packwiz.installer.target.ClientHolder
import link.infra.packwiz.installer.target.Side
import link.infra.packwiz.installer.target.path.HttpUrlPath
import link.infra.packwiz.installer.target.path.PackwizPath
import link.infra.packwiz.installer.util.delegateTransitive
import okhttp3.HttpUrl.Companion.toHttpUrl
import okio.Source
import kotlin.reflect.KType

data class ModFile(
	val name: String,
	val filename: PackwizPath<*>,
	val side: Side = Side.BOTH,
	val download: Download,
	val update: Map<String, UpdateData> = mapOf(),
	val option: Option = Option(false)
) {
	data class Download(
		val url: PackwizPath<*>?,
		val hashFormat: HashFormat<*>,
		val hash: String,
		val mode: DownloadMode = DownloadMode.URL
	) {
		companion object {
			fun mapper() = tomlMapper {
				decoder<TomlValue.String, PackwizPath<*>> { it -> HttpUrlPath(it.value.toHttpUrl()) }
				mapping<Download>("hash-format" to "hashFormat")

				delegateTransitive<HashFormat<*>>(HashFormat.mapper())
				delegate<DownloadMode>(DownloadMode.mapper())
			}
		}
	}

	@Transient
	val resolvedUpdateData = mutableMapOf<String, PackwizPath<*>>()

	data class Option(
		val optional: Boolean,
		val description: String = "",
		val defaultValue: Boolean = false
	) {
		companion object {
			fun mapper() = tomlMapper {
				mapping<Option>("default" to "defaultValue")
			}
		}
	}

	@Throws(Exception::class)
	fun getSource(clientHolder: ClientHolder): Source {
		return when (download.mode) {
			DownloadMode.URL -> {
				(download.url ?: throw Exception("No download URL provided")).source(clientHolder)
			}
			DownloadMode.CURSEFORGE -> {
				if (!resolvedUpdateData.contains("curseforge")) {
					throw Exception("Metadata file specifies CurseForge mode, but is missing metadata")
				}
				return resolvedUpdateData["curseforge"]!!.source(clientHolder)
			}
		}
	}

	@get:Throws(Exception::class)
	val hash: Hash<*>
		get() = download.hashFormat.fromString(download.hash)

	companion object {
		fun mapper(base: PackwizPath<*>) = tomlMapper {
			delegateTransitive<PackwizPath<*>>(PackwizPath.mapperRelativeTo(base))

			delegateTransitive<Option>(Option.mapper())
			delegateTransitive<Download>(Download.mapper())

			delegateTransitive<Side>(Side.mapper())

			val updateDataMapper = UpdateData.mapper()
			decoder { type: KType, it: TomlValue.Map ->
				if (type.arguments[1].type?.classifier == UpdateData::class) {
					updateDataMapper.decode<Map<String, UpdateData>>(it)
				} else {
					pass()
				}
			}
		}
	}
}